### PR TITLE
POSIX SHM new logic

### PR DESCRIPTION
**Pull request by ddurnov**
Signed-off-by: Dmitry Durnov <dmitry.durnov@intel.com>

### CODE REVIEW

**Comment by raffenet** [comment]
Looks like a good start. Running tests to get idea of correctness. test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
Looks like there's some clash between the new code and the odd/even and nolocal debug configurations, which I'm looking into. It seems to work fine in a standard configuration.

**Comment by raffenet** [comment]
I see the problem. `MPIDI_am_isend` is unconditionally calling the NM version. I'll work up a patch to check for SHM mode and call the right "mod".

**Comment by ddurnov** [comment]
That's the way I've verified most part of the logic. I had my own glue layer which used topology information. I may create a PR with the code. As we discussed with Pavan, we have decided to postpone initial "caller context" idea and go for "double checked" topology

**Comment by raffenet** [comment]
test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
I pushed a commit to fix some locality assumptions in the fbox allocation code. The branch now works for simple tests using `MPIR_CVAR_ODD_EVEN_CLIQUES=1`, but Jenkins shows regressions in more complex tests. Need to dig some more into what's going on there...

File: src/mpid/ch4/shm/posix/posix_am.h [hunk 1]
```diff
@@ -3,114 +3,401 @@
...
+                             MPI_ERR_OTHER, "**nomem", "**nomem %s", "Send Pack buffer alloc");
+
+        MPIDU_Segment_pack(segment_ptr, segment_first, &last,
+                           MPIDI_POSIX_AMREQUEST_HDR(sreq, pack_buffer));
+        MPIDU_Segment_free(segment_ptr);
+
+        send_buf = (uint8_t *) curr_sreq_hdr->pack_buffer;
+    }
+
+    const int grank = MPIDI_CH4U_rank_to_lpid(rank, comm);
```
Comment (raffenet) [inline src/mpid/ch4/shm/posix/posix_am.h]: @ddurnov I think this is another problem. We are using essentially the MPI_COMM_WORLD rank where we should be using node local rank. As it is, it works fine for single node jobs, but multinode (or debug mode) will segfault if the rank index overruns the allocated segment. @yfguo we need an efficient mechanism to translate MPI_COMM_WORLD rank to node local rank in order for this work.

**Comment by ddurnov** [comment]
Would be great if we could have a generic layer which would call SHM AM/NM AM in a way similar to the one I've added here: #2532

**Comment by raffenet** [comment]
@ddurnov I tend to agree. I think it's a cleaner solution to what we have in this branch.

**Comment by raffenet** [comment]
Plus we can probably avoid having to do multiple calculations of the local_rank if we store it in the request, knowing that it is a shm operation.

**Comment by raffenet** [comment]
@ddurnov I squashed my locality fixes down to a single commit. There's still issues with intercomm merge in it's current state. Should be reproducible without using the local debug modes. E.g. `test/mpi/coll/gather` test with 4 processes will fail.

**Comment by ddurnov** [comment]
test:jenkins/ch4/ofi

**Comment by wesbland** [comment]
@ddurnov - There's a lot of conflicts here from not having been rebased in the last 10 months. Can you guys update this to the latest master?

**Comment by wesbland** [comment]
There's one bit of bad whitespace here but it's fixed in #3034.

**Comment by wesbland** [comment]
Closing as this work is moving to #3199

**Pull request closed without merging**